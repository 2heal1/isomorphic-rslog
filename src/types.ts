export type LogLevel = 'log' | 'debug' | 'info' | 'warn' | 'error';

export type LogMessage = number | string | Error | null;

export type LogFormatter = (
  input: string | number | null | undefined,
) => string;

export interface LogType {
  label?: string;
  level: LogLevel;
  LogFormatter?: LogFormatter;
}

export type LogFunction = (message?: LogMessage, ...args: any[]) => void;

export interface Options<T extends Record<string, LogType>> {
  level?: string;
  types?: T;
}
