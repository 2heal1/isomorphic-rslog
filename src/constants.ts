export let LOG_LEVEL = {
  error: 0,
  warn: 1,
  info: 2,
  log: 3,
  verbose: 4,
} as const;
