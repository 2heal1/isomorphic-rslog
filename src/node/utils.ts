import supportsColor from 'supports-color';
import { bold } from './color';
import type { Labels, LogMessage, LogMethods, LogType } from '../types'

// https://github.com/chalk/supports-color
export const colorLevel = supportsColor.stdout ? supportsColor.stdout.level : 0;

export function getLabel(type:LogMethods,logType: LogType,labels:Labels){
    let label = '';
    if ('label' in logType) {
        label = (labels[type] || logType.label || '').padEnd(7);
        label = bold(logType.color ? logType.color(label)[0] : label)[0];
      }
  
    return [label]
  }
  
  export function finalLog(label:string[],text:string,args:string[],message?:LogMessage){
    const labelStr = label[0];
    //  gradient
    if(text && Array.isArray(message) && !(message instanceof Error)){
    console.log(`${labelStr} ${message[0]}`);
    }else{
      console.log(labelStr.length ? `${labelStr} ${text}` : text, ...args);
    }
  }