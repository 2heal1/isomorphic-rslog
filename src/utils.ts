const errorStackRegExp = /^\s*at\s.*:\d+:\d+[\s)]*$/;
const anonymousErrorStackRegExp = /^\s*at\s.*\(<anonymous>\)$/;

export const isErrorStackMessage = (message: string) =>
  errorStackRegExp.test(message) || anonymousErrorStackRegExp.test(message);

export function getLongestLabelLength(labels: (string | undefined)[]) {
  let longestLabel = '';
  labels.forEach(label => {
    if (label && label.length > longestLabel.length) {
      longestLabel = label;
    }
  });
  return longestLabel.length;
}
