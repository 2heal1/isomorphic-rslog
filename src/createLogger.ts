import { LOG_LEVEL, } from './constants';
import { isErrorStackMessage } from './utils';
import type { Options, LogMessage, Logger, LogMethods, LogTypes } from './types';

import type { gradient as TGradient } from './browser/gradient';
import type { finalLog as TFinalLog, getLabel as TGetLabel } from './browser/utils';


export let createLogger = (options: Options = {}, { getLabel, handleError, finalLog, greet, LOG_TYPES }: {
  LOG_TYPES: LogTypes;
  getLabel: typeof TGetLabel;
  finalLog: typeof TFinalLog;
  gradient: typeof TGradient,
  greet: (msg: string) => unknown,
  handleError: (msg: string) => string,

}) => {
  let maxLevel = options.level || 'log';
  const customLabels = options.labels || {};

  let log = (type: LogMethods, message?: LogMessage, ...args: string[]) => {
    if (LOG_LEVEL[LOG_TYPES[type].level] > LOG_LEVEL[maxLevel]) {
      return;
    }

    if (message === undefined || message === null) {
      return console.log();
    }

    let logType = LOG_TYPES[type];
    let text = '';

    const label = getLabel(type, logType, customLabels);

    if (message instanceof Error) {
      if (message.stack) {
        let [name, ...rest] = message.stack.split('\n');
        if (name.startsWith('Error: ')) {
          name = name.slice(7);
        }
        text = `${name}\n${handleError(rest.join('\n'))}`;
      } else {
        text = message.message;
      }
    }

    // change the color of error stacks to
    else if (logType.level === 'error' && typeof message === 'string') {

      let lines = message.split('\n');
      text = lines
        .map(line => (isErrorStackMessage(line) ? handleError(line) : line))
        .join('\n');
    } else {
      text = `${message}`;
    }

    finalLog(label, text, args, message)
  };

  let logger = {
    // greet
    greet: (message: string) => log('log', greet(message)),
  } as Logger;

  (Object.keys(LOG_TYPES) as LogMethods[]).forEach(key => {
    logger[key] = (...args) => log(key, ...args);
  });

  Object.defineProperty(logger, 'level', {
    get: () => maxLevel,
    set(val) {
      maxLevel = val;
    },
  });

  logger.override = customLogger => {
    Object.assign(logger, customLogger);
  };

  return logger;
};
