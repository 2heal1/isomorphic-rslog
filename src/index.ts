import { createLogger } from './createLogger';

export { createLogger } from './createLogger';

export const logger = createLogger();

export type {
  Options,
  LogType,
  LogLevel,
  LogMessage,
  LogFormatter,
  LogFunction,
} from './types';
