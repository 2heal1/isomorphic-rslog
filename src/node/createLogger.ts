import {createLogger as rawCreateLogger} from '../createLogger';
import  {  gray    } from './color';
import  { gradient   } from './gradient';
import  { finalLog   , getLabel   } from './utils';
import {LOG_TYPES} from './constants'
import type { Options,  } from '../types';

export function createLogger (options: Options = {},){
   return rawCreateLogger(options,{
      handleError:(msg)=>{
      const res = gray(msg);
      return Array.isArray(res) ? `${res[0]}` : `${res}`
   },
    getLabel,
    gradient,
    finalLog,
    LOG_TYPES,
    greet:(msg)=>{
     return gradient(msg)[0]
    }
   })
}