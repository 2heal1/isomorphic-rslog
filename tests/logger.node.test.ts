import { createLogger, logger } from '../src/node';
import { join } from 'path';
import { expect, test, describe, vi, Mock, beforeEach } from 'vitest';

const root = join(__dirname, '../');

expect.addSnapshotSerializer({
  test: val => typeof val === 'string' && val.includes(root),
  print: val => {
    return (val as any).toString().replaceAll(root, '<ROOT>');
  },
});

describe('logger', () => {
  beforeEach(() => {
    console.log = vi.fn();
  });

  test('should log as expected', () => {
    logger.greet(`😊 Rslog v1.0.0\n`);
    logger.log('this is a log message');
    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.success('this is a success message');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should create new logger with info level correctly', () => {
    const logger = createLogger({
      level: 'info',
    });

    logger.greet(`😊 Rslog v1.0.0\n`);
    logger.error('this is a error message');
    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.success('this is a success message');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should create new logger with warn level correctly', () => {
    const logger = createLogger({
      level: 'warn',
    });

    logger.greet(`😊 Rslog v1.0.0\n`);
    logger.error('this is a error message');
    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.success('this is a success message');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should log error with stack correctly', () => {
    console.log = vi.fn();
    const error = new Error('test error');
    logger.error(error);
    
    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toContain('error');
    expect(calls[0][0]).toContain('test error');
    expect(calls[0][0]).toContain('at ');
  });

  test('use custom labels if the value is passed', () => {
    const logger = createLogger({
      labels: {
        warn: '[ Prefix ] Warn',
        error: '[ Prefix ] Error',
        success: '[ Prefix ] Success',
        info: '[ Prefix ] Info',
        ready: '[ Prefix ] Ready',
        debug: '[ Prefix ] Debug',
      }
    });

    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.success('this is a success message');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle multiple arguments in log messages', () => {
    logger.info('Message with', 'multiple', 'arguments');
    logger.warn('Warning with', { object: 'value' });
    logger.error('Error with', 123, true);

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle empty and undefined messages', () => {
    logger.info();
    logger.warn('');
    logger.error(undefined);
    logger.debug(null);

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle start method correctly', () => {
    logger.start('Starting process');
    logger.start('Starting with', 'multiple', 'arguments');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle multiple logger instances independently', () => {
    const logger1 = createLogger({ level: 'info' });
    const logger2 = createLogger({ level: 'warn' });

    logger1.info('Info from logger1');
    logger2.info('Info from logger2'); // Should not be logged
    logger2.warn('Warning from logger2');

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle error objects with custom properties', () => {
    console.log = vi.fn();
    const customError = new Error('Custom error');
    (customError as any).customProp = 'test';
    logger.error(customError);
    
    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1);
    const output = calls[0][0].replace(/\u001b\[\d+m/g, '');
    expect(output).toMatch(/error\s+Custom error/);
    expect(output).toContain('at ');
  });

  test('should handle non-string messages', () => {
    logger.info(123);
    logger.warn(true);
    logger.error({ message: 'object message' });
    logger.debug([1, 2, 3]);

    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });
});

describe('color and formatting', () => {
  beforeEach(() => {
    console.log = vi.fn();
  });

  test('should handle nested color formatting', () => {
    const logger = createLogger();
    logger.info('\x1b[31mred\x1b[39m and \x1b[32mgreen\x1b[39m text');
    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle multiple nested styles', () => {
    const logger = createLogger();
    logger.info('\x1b[1m\x1b[31mbold red\x1b[39m\x1b[22m text');
    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle gradient text with spaces and newlines', () => {
    const logger = createLogger();
    logger.greet('Hello\nWorld with spaces');
    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle emoji in gradient text', () => {
    const logger = createLogger();
    logger.greet('Hello 👋 World 🌍');
    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });
});

describe('label formatting', () => {
  beforeEach(() => {
    console.log = vi.fn();
  });

  test('should handle empty labels', () => {
    const logger = createLogger({
      labels: {
        info: '',
        warn: '',
        error: '',
      }
    });
    
    logger.info('message');
    logger.warn('message');
    logger.error('message');
    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle very long labels', () => {
    const logger = createLogger({
      labels: {
        info: 'VeryLongLabelThatExceedsNormalLength',
        warn: 'AnotherVeryLongLabel',
      }
    });
    
    logger.info('message');
    logger.warn('message');
    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });
});

describe('error handling', () => {
  beforeEach(() => {
    console.log = vi.fn();
  });

  test('should handle circular references in error objects', () => {
    const circularObj: any = { foo: 'bar' };
    circularObj.self = circularObj;
    
    const error = new Error('Circular reference error');
    (error as any).circular = circularObj;
    
    logger.error(error);
    expect((console.log as Mock).mock.calls[0][0]).toContain('Circular reference error');
  });

  test('should handle errors with no stack trace', () => {
    const errorLike = {
      message: 'Error-like object',
      name: 'CustomError'
    };
    
    logger.error(errorLike);
    expect((console.log as Mock).mock.calls).toMatchSnapshot();
  });

  test('should handle errors with custom toString', () => {
    console.log = vi.fn();
    class CustomError extends Error {
      toString() {
        return 'Custom toString message';
      }
    }
    
    const error = new CustomError('message');
    logger.error(error);
    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1);
    const output = calls[0][0].replace(/\u001b\[\d+m/g, '');
    expect(output).toMatch(/error\s+message/);
    expect(output).toContain('at ');
  });
});

describe('log level filtering', () => {
  beforeEach(() => {
    console.log = vi.fn();
  });

  test('should respect error level', () => {
    const logger = createLogger({ level: 'error' });
    
    logger.debug('debug');
    logger.info('info');
    logger.warn('warn');
    logger.error('error');
    
    const calls = (console.log as Mock).mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toContain('error');
  });

  test('should handle invalid log level gracefully', () => {
    const logger = createLogger({ 
      // @ts-expect-error testing invalid level
      level: 'invalid' 
    });
    
    logger.info('should still work');
    expect((console.log as Mock).mock.calls.length).toBe(1);
  });
});
