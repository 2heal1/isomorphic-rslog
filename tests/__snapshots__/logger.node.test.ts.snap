// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`color and formatting > should handle emoji in gradient text 1`] = `
[
  [
    "[1m[38;2;179;250;236mH[39m[38;2;170;245;228me[39m[38;2;160;240;221ml[39m[38;2;151;235;213ml[39m[38;2;141;230;206mo[39m[38;2;141;230;206m [39m[38;2;131;224;199m👋[39m[38;2;131;224;199m [39m[38;2;122;219;191mW[39m[38;2;112;214;184mo[39m[38;2;103;209;176mr[39m[38;2;93;204;169ml[39m[38;2;84;199;161md[39m[38;2;84;199;161m [39m[38;2;74;194;154m🌍[39m[22m",
  ],
]
`;

exports[`color and formatting > should handle gradient text with spaces and newlines 1`] = `
[
  [
    "[1m[38;2;183;252;239mH[39m[38;2;178;249;234me[39m[38;2;172;246;230ml[39m[38;2;166;243;225ml[39m[38;2;160;240;221mo[39m[38;2;160;240;221m
[39m[38;2;155;237;216mW[39m[38;2;149;234;212mo[39m[38;2;143;231;207mr[39m[38;2;137;228;203ml[39m[38;2;132;224;199md[39m[38;2;132;224;199m [39m[38;2;126;221;194mw[39m[38;2;120;218;190mi[39m[38;2;114;215;185mt[39m[38;2;109;212;181mh[39m[38;2;109;212;181m [39m[38;2;103;209;176ms[39m[38;2;97;206;172mp[39m[38;2;91;203;167ma[39m[38;2;86;200;163mc[39m[38;2;80;197;158me[39m[38;2;74;194;154ms[39m[22m",
  ],
]
`;

exports[`color and formatting > should handle multiple nested styles 1`] = `
[
  [
    "[1m[36minfo   [39m[22m [1m[31mbold red[39m[22m text",
  ],
]
`;

exports[`color and formatting > should handle nested color formatting 1`] = `
[
  [
    "[1m[36minfo   [39m[22m [31mred[39m and [32mgreen[39m text",
  ],
]
`;

exports[`error handling > should handle errors with no stack trace 1`] = `
[
  [
    "[1m[31merror  [39m[22m [object Object]",
  ],
]
`;

exports[`label formatting > should handle empty labels 1`] = `
[
  [
    "[1m[36minfo   [39m[22m message",
  ],
  [
    "[1m[33mwarn   [39m[22m message",
  ],
  [
    "[1m[31merror  [39m[22m message",
  ],
]
`;

exports[`label formatting > should handle very long labels 1`] = `
[
  [
    "[1m[36mVeryLongLabelThatExceedsNormalLength[39m[22m message",
  ],
  [
    "[1m[33mAnotherVeryLongLabel[39m[22m message",
  ],
]
`;

exports[`logger > should create new logger with info level correctly 1`] = `
[
  [
    "[1m[31merror  [39m[22m this is a error message",
  ],
  [
    "[1m[36minfo   [39m[22m this is an info message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
  [
    "[1m[32mready  [39m[22m this is a ready message",
  ],
  [
    "[1m[32msuccess[39m[22m this is a success message",
  ],
]
`;

exports[`logger > should create new logger with warn level correctly 1`] = `
[
  [
    "[1m[31merror  [39m[22m this is a error message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
]
`;

exports[`logger > should handle empty and undefined messages 1`] = `
[
  [],
  [
    "[1m[33mwarn   [39m[22m ",
  ],
  [],
]
`;

exports[`logger > should handle multiple arguments in log messages 1`] = `
[
  [
    "[1m[36minfo   [39m[22m Message with",
    "multiple",
    "arguments",
  ],
  [
    "[1m[33mwarn   [39m[22m Warning with",
    {
      "object": "value",
    },
  ],
  [
    "[1m[31merror  [39m[22m Error with",
    123,
    true,
  ],
]
`;

exports[`logger > should handle multiple logger instances independently 1`] = `
[
  [
    "[1m[36minfo   [39m[22m Info from logger1",
  ],
  [
    "[1m[33mwarn   [39m[22m Warning from logger2",
  ],
]
`;

exports[`logger > should handle non-string messages 1`] = `
[
  [
    "[1m[36minfo   [39m[22m 123",
  ],
  [
    "[1m[33mwarn   [39m[22m true",
  ],
  [
    "[1m[31merror  [39m[22m [object Object]",
  ],
]
`;

exports[`logger > should handle start method correctly 1`] = `
[
  [
    "[1m[36mstart  [39m[22m Starting process",
  ],
  [
    "[1m[36mstart  [39m[22m Starting with",
    "multiple",
    "arguments",
  ],
]
`;

exports[`logger > should log as expected 1`] = `
[
  [
    "[1m[38;2;179;250;236m😊[39m[38;2;179;250;236m [39m[38;2;170;245;228mR[39m[38;2;160;240;221ms[39m[38;2;151;235;213ml[39m[38;2;141;230;206mo[39m[38;2;131;224;199mg[39m[38;2;131;224;199m [39m[38;2;122;219;191mv[39m[38;2;112;214;184m1[39m[38;2;103;209;176m.[39m[38;2;93;204;169m0[39m[38;2;84;199;161m.[39m[38;2;74;194;154m0[39m[38;2;74;194;154m
[39m[22m",
  ],
  [
    "this is a log message",
  ],
  [
    "[1m[36minfo   [39m[22m this is an info message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
  [
    "[1m[32mready  [39m[22m this is a ready message",
  ],
  [
    "[1m[32msuccess[39m[22m this is a success message",
  ],
]
`;

exports[`logger > use custom labels if the value is passed 1`] = `
[
  [
    "[1m[36m[ Prefix ] Info[39m[22m this is an info message",
  ],
  [
    "[1m[33m[ Prefix ] Warn[39m[22m this is a warn message",
  ],
  [
    "[1m[32m[ Prefix ] Ready[39m[22m this is a ready message",
  ],
  [
    "[1m[32m[ Prefix ] Success[39m[22m this is a success message",
  ],
]
`;
